src/lib/util/ioprocs.h

```diff
@@ -57,6 +57,23 @@ class read_stream
 	///   be less than or equal to the requested length.
 	/// \return An error condition if reading stopped due to an error.
 	virtual std::error_condition read(void *buffer, std::size_t length, std::size_t &actual) noexcept = 0;
+
+	/// \brief Copy from the current position in the stream
+	///
+	/// Copies the specified number of bytes from the stream to the
+	/// supplied destination stream.  May read or write less than the
+	/// requested number of bytes if the end of the stream is reached or
+	/// an error occurs.  If the stream supports seeking, reading starts
+	/// at the current position in the stream, and the current position
+	/// is incremented by the number of bytes read.  Copied data is
+	/// written by invoking the stream's write method, either once or
+	/// iteratively.
+	/// \param [in] stream Destination stream to write to.
+	/// \param [in] length Maximum number of bytes to copy.
+	/// \param [out] actual Number of bytes actually written.  Will always
+	///   be less than or equal to the requested length.
+	/// \return An error condition if reading stopped due to an error.
+	virtual std::error_condition copy(write_stream &stream, std::size_t length, std::size_t &actual) noexcept;
```

Comment:
I really dislike this on multiple counts.  In particular:
* You can’t tell whether an error came from reading or writing.
* You can’t tell how much was read but failed to be written.
* The implementation returning `errc::io_error` on reaching EOF is not intuitive.

Reply:
None of the ioprocs classes have particularly nice or well-specified ways of reporting errors, as the numerous FIXME comments elsewhere in the source reflect. The output parameters for storing the actual length read/written are particularly cumbersome and rarely tested in any rigorous way, but perhaps adding another one here would be better.

Reply:
It’s not a good idea to put something in the interface that can’t be made to work without pitfalls.  A function like this always has the potential to lose the bytes read before a failed write.  It’s just too high-level for this interface.